src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -142,6 +149,94 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
   }
 
+  // collect variables used in the expression
+  def collectVariables(expr: Expr) : Set[String] = {
+    var set = new HashSet[String]()
+    def collectVariablesInner(e: Expr) : Unit = e match {
+      case VarExpr(name) => set += name
+      case FuncExpr(function, args, agg) => args foreach (collectVariablesInner(_))
+      case BinaryOpExpr(lhs, op, rhs) => {
+        collectVariablesInner(lhs)
+        collectVariablesInner(rhs)
+      }
+      case TypecastExpr(lhs, rhs) => collectVariablesInner(lhs)
+      case _ =>
+    }
+    collectVariablesInner(expr)
+    set
+  }
+
+  def collectVariables(pattern: Pattern, definitionOnly: Boolean) : Set[String] = {
```

Comment:
I'd rather define two functions: `collectUsedVars` and `collectDefinedVars`.  Once you eliminate the `var set` from the functions, it won't feel duplicating too much.  Moreover, you can remove the `if`s and ignore the rest with `case _ =>`, right?